From cf1c18177cd83c8b9d869ae38b8e9e47d6d90283 Mon Sep 17 00:00:00 2001
From: "guoyin.chen" <guoyin.chen@freescale.com>
Date: Wed, 7 Nov 2012 16:47:36 +0800
Subject: [PATCH 4/8] ENGR00232760 Some USB Disk (disk:NPARTS=1,part:PARTN=4)
 cannot be mount

Disk uevent report the NPARTS=1, but the only partition uevent report PARTN=4
Vold will create major:-1 dev node for mount function, which will failed when
fsck_msdos to access the dev path.
Create the dev node only for those minior not be -1

Signed-off-by: guoyin.chen <guoyin.chen@freescale.com>
---
 DirectVolume.cpp | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/DirectVolume.cpp b/DirectVolume.cpp
index cc594b9..a3ddcfb 100644
--- a/DirectVolume.cpp
+++ b/DirectVolume.cpp
@@ -378,11 +378,14 @@ int DirectVolume::getDeviceNodes(dev_t *devs, int max) {
             return 1;
         }
 
-        int i;
-        for (i = 0; i < mDiskNumParts; i++) {
+        int i, j;
+        for (i = 0, j = 0; i < mDiskNumParts; i++) {
             if (i == max)
                 break;
-            devs[i] = MKDEV(mDiskMajor, mPartMinors[i]);
+            //Do not make dev for those non valid partition
+            if (mPartMinors[i] == -1)
+                continue;
+            devs[j++] = MKDEV(mDiskMajor, mPartMinors[i]);
         }
         return mDiskNumParts;
     }
-- 
1.8.0

